### PR TITLE
Add IItemHandler capability to TileMultiCrafter

### PR DESCRIPTION
**Pull request by tomKPZ**
This allows piping in/out of multiblock crafters.
Fixes https://github.com/TechReborn/RebornStorage/issues/6

### CODE REVIEW

**Comment by raoulvdberge** [comment]
Are you sure the RS pattern list is rebuild?

**Comment by modmuss50** [comment]
It should do, this isn’t how I planned to add it, as I as going to add an extra block to allow it. I will give this a good look over to make sure it works ok.

I’m also worried about the peformace issues that might arise here do to the size of the inventory being exposed.

**Comment by tomKPZ** [comment]
If you provide the texture/recipe/name for the block, I'd be glad to add it.

**Comment by tomKPZ** [comment]
Any update?  Are you planning to merge this?

**Comment by modmuss50** [comment]
What you have looks good, I was planning to add a "Storage IO"? block that you can add to the machine to access it from an external block/pipe/whatever. Texture wise I dont have a way to get one, I could try make one, or let you have a go. (Just edit the heat texture a bit?)

The block should only work on the sides (not the edges). I am also worried about performance with this. I dont see/know of an easy way to fix it?

Anyway ill be happy to merge what you have once its enabled with a block. Sorry for being a bit slow to response. If you have any questions or suggestions feel free to leave them here or on our discord.

**Comment by tomKPZ** [comment]
Done.

**Comment by modmuss50** [comment]
Texture looks fine, code looks good. Ill merge it and give it some tests before it is pushed to curse. Thanks a bunch for this.

**Comment by modmuss50** [comment]
Ok, as I expected there are some major performance issues with it. I created a 10x10x10 multiblock using the command (/rsmblock 10) and added an io port. I then had a hopper facing towards it. Before I started the profiler I placed a blank pattern into the hopper.

The profiler showed that the next tick took around 50 seconds. This is due to the hopper checking every slot (almost 40k) to see if it is valid for that item.

screen shot 2018-05-09 at 17 19 30

A much smaller multiblock (5x5) is causing the world tick to take around 500ms with a similar setup. Some optimisations could be done:

- Dont check if the pattern is valid when being inserted, this would save a large amount of time, but would allow blank patters in.

- Only allow the IO port to access a user specified page? 

- When inserting an item have the multiblock know where the next free slot is, and only allow insertion into that one slot. Something similar could be done for extraction. How ever this would not allow for blocks like interfaces to be attached to the side of the mulitblock.

- Similar to the one above, the IO port could have it own small inv and handle insertion of patters on its own in a sane way. (Unlike the hooper checking every slot)

I would like to hear your feedback on it, and see if you have any suggestions. (@raoulvdberge and @gigabit101 might have some expertise in large inv's)

TLDR:

The new Io Port is laggy.

**Comment by tomKPZ** [comment]
Thanks for the profiling and suggestions!  Is the hopper checking every slot because it wants to know if it can merge the item into an existing stack?  If so, this check is just a waste of time if the max stack size is 1.

> Dont check if the pattern is valid when being inserted, this would save a large amount of time, but would allow blank patters in.

Would the problem still exist with pipes that try to extract by item + NBT?

> Only allow the IO port to access a user specified page?

I'm strongly against this.  I feel like it defeats the purpose of having an IO port.

> When inserting an item have the multiblock know where the next free slot is, and only allow insertion into that one slot. Something similar could be done for extraction. How ever this would not allow for blocks like interfaces to be attached to the side of the mulitblock.

I feel like this is the best option, though it's slightly more difficult to implement.  The most common use case is "I just want to move a whole bunch of patterns in/out of the multiblock crafter, and I don't care which slots are used".  I can't think of a case where you'd want to insert/extract from a specific slot.

> Similar to the one above, the IO port could have it own small inv and handle insertion of patters on its own in a sane way. (Unlike the hooper checking every slot)

Is this the same as the above except instead of 1 slot, there's n>1 slots?  If my understanding is correct, I think this wouldn't be adding much.

**Comment by modmuss50** [comment]
> Would the problem still exist with pipes that try to extract by item + NBT?

Yes, it would only help the issue as it would speed up the checking to see if the item is valid for the slot. But would then allow all items into the slot. Hoppers and most other item storage devices would still need to check every single slot.

> I'm strongly against this. I feel like it defeats the purpose of having an IO port.

I agree it wouldn't be a great user experience.

>  feel like this is the best option, though it's slightly more difficult to implement. The most common use case is "I just want to move a whole bunch of patterns in/out of the multiblock crafter, and I don't care which slots are used". I can't think of a case where you'd want to insert/extract from a specific slot.

> Is this the same as the above except instead of 1 slot, there's n>1 slots? If my understanding is correct, I think this wouldn't be adding much.

These two are similar, and I agree would be the best idea, but like you said a bit of a pain to add. It would allow for adding to the multiblock as well as removing. It would also need to check for patterns on random slots every so often, this can be done slowly and per storage component. For example if a user added a pattern in the last slot the io upgrade would need to know about that slot when extracting, but when inserting it would need to know the first slot that is empty.

Another thing (Thanks @gigabit101 for the suggestion) that might be a good idea adding support for is (https://github.com/CyclopsMC/CommonCapabilitiesAPI/wiki/Slotless-Item-Handler) the item handler that colossal chests used to get over the issue. This would allow compatible mods to see the entire inventory. CommonCapabilities seems to have some other things that might be useful. I am not sure on the amount of mods that support this, so it may not be worth the effort implementing it.

**Comment by tomKPZ** [comment]
> These two are similar, and I agree would be the best idea, but like you said a bit of a pain to add. It would allow for adding to the multiblock as well as removing. It would also need to check for patterns on random slots every so often, this can be done slowly and per storage component. For example if a user added a pattern in the last slot the io upgrade would need to know about that slot when extracting, but when inserting it would need to know the first slot that is empty.

Ok, sounds like we're in agreement.  Like you said, it would probably be pretty fast if we cache the first open slot and last used slot.

Did you plan on implementing this?

> Another thing (Thanks @gigabit101 for the suggestion) that might be a good idea adding support for is (https://github.com/CyclopsMC/CommonCapabilitiesAPI/wiki/Slotless-Item-Handler) the item handler that colossal chests used to get over the issue. This would allow compatible mods to see the entire inventory. CommonCapabilities seems to have some other things that might be useful. I am not sure on the amount of mods that support this, so it may not be worth the effort implementing it.

This would certainly be the best option.  If only everyone supported it :)

**Comment by modmuss50** [comment]
Ok cool, sounds like we have a plan.

I dont think im going to have the time for at least another few weeks to sit down and work on it. I dont mind doing it but it will have to wait. If you want to have a crack at it, its up to you.

**Comment by tomKPZ** [comment]
I'll try to find some time to work on this, soon.
